lib/Yield/Calls/Slices.cpp

```diff
@@ -0,0 +1,133 @@
+/// \file Slices.cpp
+/// \brief
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include <unordered_map>
+
+#include "llvm/ADT/BreadthFirstIterator.h"
+#include "llvm/ADT/PostOrderIterator.h"
+
+#include "revng/Yield/Calls/Slices.h"
+
+using NodeView = const yield::Graph::Node *;
+
+static yield::Graph::Node *
+nodeHelper(yield::Graph &Graph, const yield::Graph::Node *Source) {
+  auto New = std::make_unique<yield::Graph::Node>(Source->data());
+  return Graph.addNode(std::move(New));
+}
+
+/// \tparam NV local `NodeView` specialization
+/// \tparam INV inverted location `NodeView` specialization
+template<typename NV, typename INV>
+yield::Graph
+preservingSliceImpl(const yield::Graph &Input, const MetaAddress &SlicePoint) {
+  auto LookupHelper = [&SlicePoint](NodeView Node) {
+    return Node->Address == SlicePoint;
+  };
+  auto Entry = llvm::find_if(Input.nodes(), LookupHelper);
+  revng_assert(Entry != Input.nodes().end());
+
+  // Find the rank of each node, such that for any node its rank is equal to
+  // the highest rank among its children plus one.
+  std::unordered_map<const yield::Graph::Node *, size_t> Ranks;
+  std::vector<NodeView> PostOrder(llvm::po_begin(NV{ *Entry }),
+                                  llvm::po_end(NV{ *Entry }));
+  for (NodeView CurrentNode : llvm::reverse(PostOrder)) {
```

Comment:
Why not the following?
```c++
for (NodeView CurrentNode : llvm::ReversePostOrderTraversal(Entry)) {
}
```
It's more concise and does exactly the same thing.

Reply:
I don't remember the specifics, but I was running into some trouble with `ReversePostOrderTraversal` when passing an `llvm::Inverse` node into it. If you want, I can investigate it again, but there are cases when `llvm::reverse(llvm::post_order(llvm::Inverse<T>(...)))` behave differently from `llvm::ReversePostOrderTraversal(llvm::Inverse<T>(...))`.

Reply:
This shouldn't happen. If it is different something is wacky is going on with out graph. So yes, please, investigate.

Reply:
I'm closing this thread as per the discussion in mattermost.